include/aws/s3/s3_buffer_pool.h

```diff
@@ -0,0 +1,136 @@
+};
+
+/**
+ * Polymorphic ticket.
+ */
+struct aws_s3_buffer_ticket {
+    struct aws_s3_buffer_ticket_vtable *vtable;
+    struct aws_ref_count ref_count;
+    void *impl;
+};
+
+AWS_S3_API struct aws_byte_buf aws_s3_buffer_ticket_claim(struct aws_s3_buffer_ticket *ticket);
+
+AWS_S3_API struct aws_s3_buffer_ticket *aws_s3_buffer_ticket_acquire(struct aws_s3_buffer_ticket *ticket);
+AWS_S3_API struct aws_s3_buffer_ticket *aws_s3_buffer_ticket_release(struct aws_s3_buffer_ticket *ticket);
+
+struct aws_s3_buffer_pool;
+
+struct aws_s3_buffer_pool_vtable {
+    /* Reserve a ticket. Returns a future that is granted whenever reservation can be made. */
+    struct aws_future_s3_buffer_ticket *(
+        *reserve)(struct aws_s3_buffer_pool *pool, struct aws_s3_buffer_pool_reserve_meta meta);
+
+    /**
+     * Trim the pool. This is mostly a suggestion, which pool can decide to ignore. Triggered by CRT when
+     * client has been idle for some time.
+     **/
+    void (*trim)(struct aws_s3_buffer_pool *pool);
+
+    /* Implement below for custom ref count behavior. Alternatively set those to null and init the ref count. */
+    void (*acquire)(struct aws_s3_buffer_pool *pool);
+    void (*release)(struct aws_s3_buffer_pool *pool);
+};
+
+/**
+ * Polymorphic buffer pool.
+ */
+struct aws_s3_buffer_pool {
+    struct aws_s3_buffer_pool_vtable *vtable;
+    struct aws_ref_count ref_count;
+    void *impl;
+};
+
+AWS_S3_API struct aws_future_s3_buffer_ticket *aws_s3_buffer_pool_reserve(
+    struct aws_s3_buffer_pool *buffer_pool,
+    struct aws_s3_buffer_pool_reserve_meta meta);
+AWS_S3_API void aws_s3_buffer_pool_trim(struct aws_s3_buffer_pool *buffer_pool);
+
+AWS_S3_API struct aws_s3_buffer_pool *aws_s3_buffer_pool_acquire(struct aws_s3_buffer_pool *buffer_pool);
+AWS_S3_API struct aws_s3_buffer_pool *aws_s3_buffer_pool_release(struct aws_s3_buffer_pool *buffer_pool);
+
+/**
+ * Buffer pool configuration options.
+ */
+struct aws_s3_buffer_pool_config {
+    struct aws_s3_client *client; /* Client creating the pool. */
+    size_t part_size;             /* Default part size of the client. */
+    size_t max_part_size;         /* Max part size configured on the client. */
+    size_t memory_limit;          /* Memory limit set on the client. */
+};
+
+/**
+ * Factory to construct the pool for the given config. Passes along buffer related info configured on the client, which
+ * factory may ignore when considering how to construct pool.
+ * This implementation should fail if pool cannot be constructed for some reason (ex. if config params cannot be met),
+ * by logging failure reason, returning null and raising aws_error.
+ */
+typedef struct aws_s3_buffer_pool *(aws_s3_buffer_pool_factory_fn)(struct aws_allocator *allocator,
+                                                                   struct aws_s3_buffer_pool_config config);
+
+AWS_EXTERN_C_END
+AWS_POP_SANE_WARNING_LEVEL
```

Comment:
The comment describing `aws_s3_buffer_pool_factory_fn` is clear, but consider adding a brief explanation of how consumers should implement this function to match expectations of the S3 client.